### ISSUE TITLE
Use a different operator for time arithmatic

### ISSUE THREAD

**Issue by nathanielc** [influxdata/flux#2026]
Currently the SPEC defines  addition of times and durations using the `+` operator.

We note in the SPEC that the `+` for time and durations is left associative and does not commute. Maybe we should use a different operator since `+` normally communicates that its operands can associate and commute?

My proposal is to use the `@` operator. 

The examples in the SPEC would become:

```
2018-01-01T00:00:00Z @ 1d       // 2018-01-02T00:00:00Z
2018-01-01T00:00:00Z @ 1mo      // 2018-02-01T00:00:00Z
2018-01-01T00:00:00Z @ 2mo      // 2018-03-01T00:00:00Z
2018-01-31T00:00:00Z @ 2mo      // 2018-03-31T00:00:00Z
2018-02-28T00:00:00Z @ 2mo      // 2018-04-28T00:00:00Z
2018-01-31T00:00:00Z @ 1mo      // 2018-02-28T00:00:00Z, February 31th is rolled back to the last day of the month, February 28th in 2018.

// Addition and subtraction of durations to date times does not commute
2018-02-28T00:00:00Z @ 1mo @ 1d // 2018-03-29T00:00:00Z
2018-02-28T00:00:00Z @ 1d @ 1mo // 2018-04-01T00:00:00Z
2018-01-01T00:00:00Z @ 2mo @ -1d // 2018-02-28T00:00:00Z
2018-01-01T00:00:00Z @ -1d @ 3mo // 2018-03-31T00:00:00Z

// Addition and subtraction of durations to date times applies months, days and seconds in that order.
2018-01-28T00:00:00Z @ 1mo @ 2d // 2018-03-02T00:00:00Z
2018-01-28T00:00:00Z @ 1mo2d    // 2018-03-02T00:00:00Z
2018-01-28T00:00:00Z @ 2d @ 1mo // 2018-02-28T00:00:00Z, explicit left associative add of 2d first changes the result
2018-02-01T00:00:00Z @ 2mo2d    // 2018-04-03T00:00:00Z
2018-01-01T00:00:00Z @ 1mo30d   // 2018-03-02T00:00:00Z, Months are applied first to get February 1st, then days are added resulting in March 2 in 2018.
2018-01-31T00:00:00Z @ 1mo1d    // 2018-03-01T00:00:00Z, Months are applied first to get February 28th, then days are added resulting in March 1 in 2018.
```

I read like `time at duration`. Using multiple `@` should be discouraged in general and I think the syntax makes that more clear. Using multiple `@` is different from using a single `@` with a merged duration, this might provide the right syntactic pressure to write correct code.

I am not sure I want yet another operator, but it would make it clear that it is something special and doesn't operate based on existing invalid assumptions of the `+` operator. 

As an aside it would make the type inference bits easier as well, but this is just a convenience for us as developers of Flux it shouldn't inform the user experience.

**Comment by jsternberg** [comment]
I'm not sure this would clear anything up. This reminds me of the debate about string concatenation in languages. Since `+` is commutative, anything where you use `+` to combine strings is considered automatically wrong. I don't necessarily believe that, but it is a fairly reasonable belief in my own opinion.

So I think there are two components to it. Should `+` always be commutative and, if it should be, what should be the alternative?

I don't know if I agree that `+` should always be commutative. Maybe in math it would be, but I'd like to make it consistent and ensure that strings then follow the same rule. We could do that fairly easily with string expressions though so I don't have a strong preference on this.

I don't think `@` is very clear. I read it as `at` and it doesn't make sense to me what `time at duration` means. It definitely doesn't imply to me that it is addition. I'm interested in potential other operators that may make sense, but `+` does have the advantage that it automatically implies advancing. The only other symbol I can think of that may work is `>>` and `<<` to imply that you are shifting the time by that amount.

So if we decide that `+` should always be commutative, maybe `>>` and `<<` would be better symbols?

**Comment by wolffcm** [comment]
If I had to just look at `2018-01-01T00:00:00Z @ 1d` without any context, I'm not sure that I would be able to understand what it meant.  The meaning of `+` is super clear and intuitive.  I do like that `@` feels less symmetrical, though.  What about `@+`?

**Comment by affo** [comment]
I agree with @jsternberg, given that `+` is not commutative in at least another case (string concat, which we support) I think we could keep that.

BTW, I think @jsternberg proposal for shifting `>>` `<<` is more intuitive and matches the concept of shifting. Plus, we could reuse it in the future for shifting arrays.

**Comment by nathanielc** [comment]
What about `++` its still reads as addition but its clearly different? Apparently Elm use `++` to mean the `appendable`  type class. So they use it for string concat and List append etc. Not exactly the same thing as time duration addition, but an interesting idea nonetheless.

**Comment by affo** [comment]
@nathanielc Yeah, it seems interesting! Aren't you worried about confusion of the `++` in C-like languages?

I still prefer the "shift" `>>` idea 😄

**Comment by jsternberg** [comment]
It seems if we create any operator where the two sides are going to be different types, then it will be one operator per type from what I understand.

I'm wondering if we should just abandon a time arithmetic operator and instead just use a library.

```
import "date"
date.add(t: 2019-01-01T00:00:00Z, d: 1mo)
```

It is simple and tells us exactly what is happening. It also matches better with the current timezone proposal because we can add a `loc` parameter fairly easily. At the moment, the functions where you would need to do this kind of time math either have the math done on the client side and are inserted as a time or happen automatically by inserting a duration. I am wondering if we really need a special operator for time math.

**Comment by poulhs** [comment]
I surely prefer the `+`.
Things like
```
2018-01-01T00:00:00Z + 1d               // 2018-01-02T00:00:00Z
2018-01-01T00:00:00Z + 1d + 1h       // 2018-01-02T01:00:00Z
2018-01-01T00:00:00Z + 1d - 1h       // 2018-01-01T23:00:00Z
```
sounds natural when reading. And `+` is used by other languages when handling dates and durations.
And no-one expects  `"con" + "cat"` to be commutative, so why should they think so with dates.

Besides, in computing `+` isn't _really_ commutative (corner cases when you f.x. add enough small  numbers to a large enough number)

I think `+` sounds and looks betters than both `@` and `>>`
